### ISSUE TITLE
False IOException: Allocation decider fails 5 times, but showing a wrong root cause

### ISSUE THREAD

**Issue by gmoskovicz** [elastic/elasticsearch#22968]
When performing a restart, if the allocation fails after `index.allocation.max_retries` it will be shown in the allocation explanation:

```
      "decisions" : [
        {
          "decider" : "max_retry",
          "decision" : "NO",
          "explanation" : "shard has already failed allocating [5] times vs. [5] retries allowed unassigned_info[[reason=ALLOCATION_FAILED], at[2017-02-03T13:08:47.223Z], failed_attempts[5], delayed=false, details[failed to create shard, failure IOException[No space left on device]], allocation_status[no_attempt]] - manually call [/_cluster/reroute?retry_failed=true] to retry"
        },
....
....
        {
          "decider" : "awareness",
          "decision" : "NO",
          "explanation" : "there are too many shards on the node for attribute [xxxxxxxxx], there are [2] total shards for the index  and [3] total attributes values, expected the node count [2] to be lower or equal to the required number of shards per attribute [1] plus leftover [0]"
        }
```

In this example you can see why the decision was NO. Other decision makers where YES. In this case we had two reasons:

1. The awareness couldn't be satisfied based on the current number of nodes.
2. It tried 5 times and didn't succeed.

So far, the main cause is shown as a IOException issue:

```
  "unassigned_info" : {
    "reason" : "ALLOCATION_FAILED",
    "at" : "2017-02-03T13:08:47.223Z",
    "failed_attempts" : 5,
    "delayed" : false,
    "details" : "failed to create shard, failure IOException[No space left on device]",
    "allocation_status" : "no_attempt"
  },
```

There was plenty of disks space in the nodes, and we confirmed that. Moreover the decision around this was YES:

```
        {
          "decider" : "disk_threshold",
          "decision" : "YES",
          "explanation" : "enough disk for shard on node, free: [64.7gb], shard size: [0b], free after allocating shard: [64.7gb]"
        },
```

Attached full example: [example.txt](https://github.com/elastic/elasticsearch/files/751404/example.txt)

**Comment by ywelsch** [comment]
You're confusing two things:
- the reason why the shard became unassigned. This is reflected in unassigned_info, as explained in the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-allocation-explain.html).
- the reason why the shard is still unassigned at this very moment. This is shown by the node decisions.

Reading the output correctly, this means that the shard failed to be allocated 5 times in a row (i.e. it was assigned to a node, then initialization on a node failed, so it went back to unassigned, repeated 5 times). The fifth time initialization on a node failed was because of the IOException[No space left on device]. At that point, the master stopped trying to allocate this shard due to the maximum number of retries. Now, at the time of executing the explain API, the shard cannot be allocated because of the number of retries as well as awareness constraints (maybe another node has dropped out of the cluster since the last retry).

I wonder how we can improve the explain output here to avoid the confusion. We can expand the explanation for the `max_retry` decider to make it clear that the shown unassigned_info represents the reason the shard became unassigned. As for clarifying the top-level unassigned_info, I'm not sure how to avoid the confusion except by removing it altogether (which would be a shame as it adds so much more context).

**Comment by bleskes** [comment]
Looking at [MaxRetryAllocationDecider](https://github.com/elastic/elasticsearch/blob/5.2/core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/MaxRetryAllocationDecider.java#L71), I see we already output the unassigned info there:

```
                decision = allocation.decision(Decision.NO, NAME, "shard has exceeded the maximum number of retries [%d] on " +
                    "failed allocation attempts - manually call [/_cluster/reroute?retry_failed=true] to retry, [%s]",
                    maxRetry, unassignedInfo.toString());
```

so, I'm not sure that output above is complete. That said, I think an explicit "last failure: [%s]" prefix for the unassigned info will nice.

**Comment by abeyad** [comment]
> That said, I think an explicit "last failure: [%s]" prefix for the unassigned info will nice.

I think this is the best and simplest way to provide clarity.  

> Looking at MaxRetryAllocationDecider, I see we already output the unassigned info there:

The output above does indeed have the unassigned info in the explanation for the max retry decider.

**Comment by gmoskovicz** [comment]
Thanks for the ideas here. What i want to point out is that:

- There was plenty of disk space (over 90% free), and so far,  `No space left on device` was shown.
- I expect that the 5 failures where related only to the `awareness` decider.
- I confirmed that after forcing an empty reroute, everything started working again because the `awareness` succeeded after one more node (with another awareness setting) was added.
